File: grpc-pg-bench/rust-tokio/src/service.rs
//! The `bench.v1.CommandService` gRPC implementation. This layer is thin on
//! purpose: compute the checksum, stamp the receive time, delegate to `Db`,
//! and map the result back onto the proto response. All database concerns live
//! in `db.rs`; all wire concerns are these three methods.

use std::time::{SystemTime, UNIX_EPOCH};

use tonic::{Request, Response, Status};

use crate::db::Db;
use crate::fnv::fnv1a_32;
use crate::proto::bench_v1::command_service_server::CommandService;
use crate::proto::bench_v1::{CommandRequest, CommandResponse, GetStateRequest, StateResponse};

pub struct CommandSvc {
    db: Db,
}

impl CommandSvc {
    pub fn new(db: Db) -> Self {
        Self { db }
    }
}

#[tonic::async_trait]
impl CommandService for CommandSvc {
    /// Single autocommit INSERT.
    async fn execute(
        &self,
        req: Request<CommandRequest>,
    ) -> Result<Response<CommandResponse>, Status> {
        let recv = now_micros();
        let r = req.into_inner();
        let checksum = fnv1a_32(&r.payload);

        let id = self
            .db
            .insert_command(&r.workflow_id, &r.command_type, &r.payload, r.seq, checksum as i64)
            .await?;

        Ok(Response::new(CommandResponse {
            id,
            checksum,
            received_at_micros: recv,
        }))
    }

    /// Three statements in one transaction (command + state + outbox).
    async fn execute_tx(
        &self,
        req: Request<CommandRequest>,
    ) -> Result<Response<CommandResponse>, Status> {
        let recv = now_micros();
        let r = req.into_inner();
        let checksum = fnv1a_32(&r.payload);

        let id = self
            .db
            .execute_tx(&r.workflow_id, &r.command_type, &r.payload, r.seq, checksum as i64)
            .await?;

        Ok(Response::new(CommandResponse {
            id,
            checksum,
            received_at_micros: recv,
        }))
    }

    /// Single read by workflow_id.
    async fn get_state(
        &self,
        req: Request<GetStateRequest>,
    ) -> Result<Response<StateResponse>, Status> {
        let workflow_id = req.into_inner().workflow_id;

        let resp = match self.db.get_state(&workflow_id).await? {
            Some(s) => StateResponse {
                found: true,
                workflow_id,
                state: s.state,
                version: s.version,
                updated_at_micros: s.updated_at_micros,
            },
            None => StateResponse {
                found: false,
                workflow_id,
                state: String::new(),
                version: 0,
                updated_at_micros: 0,
            },
        };
        Ok(Response::new(resp))
    }
}

/// Server receive timestamp in unix micros. Saturates to 0 if the clock is
/// somehow before the epoch — same defensive default as the Go stack.
#[inline]
fn now_micros() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_micros() as i64)
        .unwrap_or(0)
}

File: grpc-pg-bench/go-gorm/service.go
package main

import (
	"context"
	"time"

	benchv1 "github.com/beam/grpc-pg-bench/gen/benchv1"
)

// commandService implements bench.v1.CommandService. Thin on purpose: compute
// the checksum, stamp receive time, delegate to the GORM Db, map the result
// back onto the proto response. All DB concerns live in db.go.
type commandService struct {
	benchv1.UnimplementedCommandServiceServer
	db *Db
}

// Execute: single autocommit INSERT (through GORM Create).
func (s *commandService) Execute(ctx context.Context, req *benchv1.CommandRequest) (*benchv1.CommandResponse, error) {
	recv := time.Now().UnixMicro()
	checksum := fnv1a(req.Payload)

	id, err := s.db.InsertCommand(ctx, req.WorkflowId, req.CommandType, req.Payload, req.Seq, int64(checksum))
	if err != nil {
		return nil, err
	}
	return &benchv1.CommandResponse{
		Id:               id,
		Checksum:         checksum,
		ReceivedAtMicros: recv,
	}, nil
}

// ExecuteTx: three statements in one transaction (command + state + outbox).
func (s *commandService) ExecuteTx(ctx context.Context, req *benchv1.CommandRequest) (*benchv1.CommandResponse, error) {
	recv := time.Now().UnixMicro()
	checksum := fnv1a(req.Payload)

	id, err := s.db.ExecuteTx(ctx, req.WorkflowId, req.CommandType, req.Payload, req.Seq, int64(checksum))
	if err != nil {
		return nil, err
	}
	return &benchv1.CommandResponse{
		Id:               id,
		Checksum:         checksum,
		ReceivedAtMicros: recv,
	}, nil
}

// GetState: single read by workflow_id.
func (s *commandService) GetState(ctx context.Context, req *benchv1.GetStateRequest) (*benchv1.StateResponse, error) {
	row, err := s.db.GetState(ctx, req.WorkflowId)
	if err != nil {
		return nil, err
	}
	if row == nil {
		return &benchv1.StateResponse{Found: false, WorkflowId: req.WorkflowId}, nil
	}
	return &benchv1.StateResponse{
		Found:           true,
		WorkflowId:      row.WorkflowID,
		State:           row.State,
		Version:         row.Version,
		UpdatedAtMicros: row.UpdatedAtMicros,
	}, nil
}
